Addition/equality operators pascal cpp python

NoahJCross opened the pull request:
# Description
Added generation of equality (==) and inequality (!=) operators to the ERB templates for SplashKit types. This ensures consistent functionality across ffi bindings and reduces manual coding effort.

Special handling was added for color comparisons to use a small epsilon (0.004) when comparing floating-point RGBA values, ensuring consistent color equality checks across all languages.

This allows you to check equality of these structs. for example color_black() == color_black()

# Changes
- Updated the ERB template to generate:
    - C++: inline operator== and operator!=
    - C#: operator == and operator != with proper Equals and GetHashCode overrides
    - Pascal: operator = and operator <>
    - Python: __eq__ and __ne__ methods

# Example Generated Code
```cpp
inline bool operator==(const point_2d& lhs, const point_2d& rhs) {
    return 
        lhs.x == rhs.x && 
        lhs.y == rhs.y; 
}

inline bool operator!=(const point_2d& lhs, const point_2d& rhs) {
    return !(lhs == rhs);
}
```
# Example Usage
```cpp
point_2d p1 = {1.0, 2.0};
point_2d p2 = {1.0, 2.0};
if (p1 == p2)
```

Testing
- [x] Verified generation works for all struct types
- [x] Confirmed proper formatting and indentation
- [x] Tested generated operators function correctly

NoahJCross committed:
Added equality for cpp, pascal, and python

NoahJCross committed:
Added equality for colors

NoahJCross committed:
reverted c_bool change

NoahJCross committed:
restored rust file

NoahJCross committed:
fixed casing for csharp

NoahJCross committed:
Added operators file

NoahJCross committed:
Added operators file

Eswar-S-Sethu commented on res/translators/cpp/interface/types/struct.h.erb:
```diff
@@ -19,6 +19,22 @@ typedef struct {
 %>
 } <%= struct[:name] %>;
+
+inline bool operator==(const <%= struct[:name] %>& lhs, const <%= struct[:name] %>& rhs) {
```
Comment: Overloading the equality operator (==) is essential for comparing objects of this type. The use of inline suggests this operator is intended for use in headers, which is appropriate.
Suggestion: Ensure struct[:name] is sanitized or validated in the template to prevent unexpected outputs.

Eswar-S-Sethu commented on res/translators/cpp/interface/types/struct.h.erb:
```diff
@@ -19,6 +19,22 @@ typedef struct {
+<%  end %>
+}
+
+inline bool operator!=(const <%= struct[:name] %>& lhs, const <%= struct[:name] %>& rhs) {
```
Comment: Overloading the inequality operator (!=) by relying on the equality operator (==) is a clean and concise approach.

Eswar-S-Sethu commented on res/translators/cpp/interface/types/struct.h.erb:
```diff
@@ -19,6 +19,22 @@ typedef struct {
+}
+
+inline bool operator!=(const <%= struct[:name] %>& lhs, const <%= struct[:name] %>& rhs) {
+    return !(lhs == rhs);
```
Comment: This is straightforward and leverages the already implemented operator==. No issues here.
Suggestion: If performance is critical and operator== involves many fields, a custom implementation for operator != might sometimes be more efficient.

Eswar-S-Sethu commented on res/translators/csharp/SplashKit/classes/structs.cs.erb:
```diff
@@ -47,6 +47,37 @@
+
+    public override bool Equals(object? obj)
+    {
+        if (obj is <%=struct[:name].to_pascal_case()%>)
```
Consider adding a null check before this line to ensure obj is not null and improve robustness.

NoahJCross replied:
csharp is operator handles null already.

Eswar-S-Sethu commented on res/translators/csharp/SplashKit/classes/structs.cs.erb:
```diff
@@ -47,6 +47,37 @@
+        return Math.Abs(lhs.R - rhs.R) < 0.004 &&
+               Math.Abs(lhs.G - rhs.G) < 0.004 &&
+               Math.Abs(lhs.B - rhs.G) < 0.004 &&
+               Math.Abs(lhs.A - rhs.A) < 0.004;
```
You could refactor this repeated logic into a helper method, e.g., AreClose(float a, float b). This would make the code cleaner and easier to maintain.
The hardcoded tolerance value 0.004 appears multiple times. Consider defining it as a const float Tolerance for easier adjustments in the future.

Eswar-S-Sethu commented on res/translators/pascal/implementation/operators.pas.erb:
```diff
@@ -0,0 +1,26 @@
+<%
+  @structs.each do |struct|
+%>
+operator = (const left, right: <%= struct[:name].type_case %>): Boolean;
```
Consider adding a comment explaining the purpose of this operator overload. It would help future maintainers understand its intent, especially if the Boolean equality logic has specific use cases.

Eswar-S-Sethu commented on res/translators/pascal/implementation/operators.pas.erb:
```diff
@@ -0,0 +1,26 @@
+<%  struct[:fields].each_with_index do |(field_name, field_data), idx|
+      is_last = idx == struct[:fields].length - 1
+      field_name = field_name.field_case
+      sk_comparison_for(field_name, field_data, is_last).each do |comparison| %>
```
This loop is dynamically generating comparisons, which is great for flexibility. However, adding inline comments or documentation for sk_comparison_for would help clarify how each comparison is constructed.

Eswar-S-Sethu commented on res/translators/pascal/implementation/operators.pas.erb:
```diff
@@ -0,0 +1,26 @@
+
+operator <> (const left, right: <%= struct[:name].type_case %>): Boolean;
+begin
+  result := not (left = right);
```
The use of not is clear here. However, adding a brief explanation about why it directly negates the equality operator could provide more clarity for maintainers.

Eswar-S-Sethu commented on src/translators/cpp.rb:
```diff
@@ -166,5 +168,21 @@ def sk_update_fn_for(function)
+
+    def comparison_statement(field_name, is_last)
+      if %w[r g b a].include?(field_name)
+        "abs(lhs.#{field_name} - rhs.#{field_name}) < 0.004#{is_last ? ';' : ' &&'}"
```
The tolerance value 0.004 is hardcoded. If this tolerance is subject to change or used in multiple places, define it as a constant for maintainability.

NoahJCross replied:
0.004 is a fixed tolerance specifically calculated for the 1.0 to 255 color value conversion. It's unlikely to ever need changing

Eswar-S-Sethu reviewed (commented):
The code is well-structured and functional, but there are minor improvements to consider:
1. Simplify and document logic where possible for clarity and future-proofing.
2. Add input validation to ensure robustness.
With these adjustments, the implementation will be more maintainable and easier to extend.

NoahJCross committed:
Fixed csharp color equality

WhyPenguins committed:
Revert 'translate' symlink removal

WhyPenguins committed:
small adjustments
 - seems that `abs` is a minefield in C++...
   https://stackoverflow.com/a/21393170
